MARP-2578 support for internationalization of the ivy cockpit

phhung-axonivy opened the pull request:
Hi @weissreto 
We have applied CMS for the cockpit.
Could you help me check and process the next step?

ivy-lli commented:
Hi @phhung-axonivy 
Thanks for your work; however, I'm not sure why you chose this strategy of doing one big PR that changes everything at once. 
The cockpit is under active development, and it is nearly impossible to get a window where this PR has no conflicts and all tests will run. Also, it is much harder to review such a huge PR...

If possible, I would highly recommend splitting this into small chunks of PR's:
- starting with the first PR containing the cms files, which e.g. translates only the dashboard
- followed by PR's for every single page or chunks of pages (features or sections or whatever)
This makes the work for you easier to hold the PR's without merge conflicts and with all tests running successfully
And for us, it is a lot easier to review the PR's.

weissreto commented:
Hi @phhung-axonivy 

I think your branch is outdated that is why the build is failing. Can you rebase you branch on the current master branch?
I will then try to review it and merge it back to master.

Thanks for your work
Regards
Reto

phhung-axonivy commented:
Hi @weissreto @ivy-lli 
I checked the build result. There are some EnvironmentNotAvailable exceptions.
<img width="695" height="176" alt="image" src="https://github.com/user-attachments/assets/36fa2f39-ee00-4c5c-8be8-59be6da73a40" />
In the test, Ivy.cm().co() or Ivy.cms().co() don't work correctly.
<img width="825" height="198" alt="image" src="https://github.com/user-attachments/assets/313f3897-71a7-4747-98ef-f84c239aa351" />

How can I resolve this problem? e.g. adding configuration?

phhung-axonivy commented:
Hi @ivy-lli @weissreto 
I have replaced all Ivy.cms().co() to Ivy.cm.content(url).replace(name, value).get().
I have added the annotation @IvyTest in test classes and fixed some bugs related to the CMS.

On the designer, after I added @IvyTest, there are some failed tests:
```
TestHealthBean.messages()
TestHealthBean.severity()
TestHealthBean.checkNow()
TestSystemDatabaseMonitorBean.transactionsMonitor()
TestSessionMonitorBean.sessionsMonitor()
All test methods of the class TestProcessExecutionBean
```
There are some failed tests. Does the annotation @IvyTest not work correctly on the Jenkins build?
<img width="1249" height="497" alt="image" src="https://github.com/user-attachments/assets/0c64b723-0c80-4e11-8af5-3d56c00821d7" />

For fixing failed tests, what do you think I should do?

phhung-axonivy commented on engine-cockpit/cms/cms_en.yaml:
```diff
@@ -0,0 +1,1493 @@
+  LogInWith: Log in with
+  LoginFooterCopyright: Copyright Ⓒ Axon Ivy
+  LoginFooterEngineCockpit: Engine Cockpit
+  NoAdministratorsConfigured: 'No administrators configured, check your '
```
Because the after is an HTML tag, I will split this CMS into two parts.
<img width="743" height="32" alt="image" src="https://github.com/user-attachments/assets/21359617-706e-45a8-b0f6-776ac77bfc0d" />
->
<img width="746" height="25" alt="image" src="https://github.com/user-attachments/assets/036e4981-1218-451b-8881-f4cfc08055e0" />
Do you have any suggestion?

ivy-lli replied:
Ah, then I think this is fine. 
Don't get me wrong, I've just marked all of this (at least those I've seen). If they can be improved great if it doesn't make much sense then it is also fine for me. 
As always in such cases maybe change the sentence would be also fine for me to e.g. `No administrators configured, check out the documentation.` and after that the link. But this is also good.

phhung-axonivy commented on engine-cockpit/webContent/includes/components/application/information.xhtml:
```diff
@@ -8,26 +8,26 @@
+        <p:commandButton id="showDeployment" title="#{ivy.cm.co('/common/Deployment')}" value="#{ivy.cm.co('/common/Deployment')}"
           style="margin-top: 20px; min-width: 150px;" type="button" icon="si si-upload-bottom"
           onclick="PF('fileUploadModal').show()" disabled="#{!uploadHelperBean.deploymentPossible}" />
         <p:tooltip for="deployBtnHint" rendered="#{!uploadHelperBean.deploymentPossible}" value="#{uploadHelperBean.deploymentPossibleReason}" escape="false" /> <!-- contains no user input, contains html-->
```
I'm sorry, I can not get your idea.

ivy-lli replied:
I know it's not your fault; this was like this before. 
The tooltip also has an `escape="false"` on it with a comment that the content of it does not contain user data, so it is not dangerous.
However, now the content is coming from the cms, and if the translator makes a mistake or wants to inject malicious code, it is much easier now. There should not be any HTML stuff inside the CMS, and there should not be any escape="false" widget which shows cms entries. 
Even if the translations are going through our review process, everyone needs to check for such things, and it is very likely that a mistake will happen.

ivy-lli commented on engine-cockpit/cms/cms_en.yaml:
```diff
@@ -0,0 +1,1493 @@
+  ViewFile: View File
+connectionTestResult:
+  ConnectToDatabaseFailMessage: 'An error occurred: %s'
+  ConnectToDatabaseSuccessMessage: 'Successfully connected to database: %s %s with JDBC Version %s'
```
I think it is because you run the test inside of another context where the Ivy.cm is no longer available. However I think I've written in another comment how you can use never the less inside of this testConnection. When you save the Ivy.cm() first in a variable and add this as parameter to the testConnection, then the evaluation for the ConentManagementSystem should be done before the context is different. (Didn't tried it but maybe worth to try for you).

ivy-lli reviewed (commented):
some more feedback

phhung-axonivy commented:
Hi @ivy-lli 
I have already handled your feedback.
However, some web tests still failed.
Could you support me?

ivy-lli reviewed (commented):
Cool thanks for your work, the feedback looks good.
I did one last check and found one more problematic escape="false". see blow.

Yes I will have a look at the WebTests and check why they are failing

ivy-lli reviewed (approved):
Yey the webtests are now also good! I needed to fix some wrong labels to achieve that :)
Hopefully there are no other labels that are wrong, but I think we need to do some smoke testing anyways.
Again thanks for your work, I will merge this PR 🚀
